addons/account/models/account_account.py

```diff
@@ -864,6 +1018,603 @@ def get_import_templates(self):
+                        lock_date=user_lock_date,
+                        locked_account_names=", ".join(account.display_name for account in locked_accounts),
+                    ))
+            else:
+                code_by_company[company] = accounts[0].with_company(company).code
+
+        return account_to_merge_into or self[0], code_by_company
+
+    def _action_merge_get_user_confirmation(self, account_to_merge_into):
+        """ Open a RedirectWarning asking the user whether to proceed with the merge. """
+        if self.env.context.get('account_merge_confirm'):
+            return
+
+        is_irreversible = len(self.company_ids) != sum(len(account.company_ids) for account in self)
+        accounts_to_remove = self - account_to_merge_into
+
+        msg = _("Are you sure? This will perform the following operations:\n")
+        for account in accounts_to_remove:
+            msg += _(
+                "- %(account_1)s (company: %(companies_1)s) will be merged into %(account_2)s (company: %(companies_2)s)\n",
+                account_1=account.with_company(account.company_ids[:1]).display_name,
+                companies_1=",".join(account.company_ids.mapped('name')),
+                account_2=account_to_merge_into.with_company(account_to_merge_into.company_ids[:1]).display_name,
+                companies_2=",".join(account_to_merge_into.company_ids.mapped('name')),
+            )
+        if is_irreversible:
+            msg += _(
+                "This cannot be undone because you are merging accounts belonging to the same company.\n"
+                "After merging, we won't be able to separate journal items based on which account they originally referenced."
+            )
+        action = self.env['ir.actions.actions']._for_xml_id('account.action_merge_accounts')
+        raise RedirectWarning(msg, action, _("Merge"), additional_context={**self.env.context, 'account_merge_confirm': True})
+
+    def _action_merge(self, account_to_merge_into, code_by_company):
+        """ Perform the merge of the accounts in `self` into `account_to_merge_into`.
+        This will update the account codes of `account_to_merge_into` based on those of `self`,
+        and update keys in DB from `self` to `account_to_merge_into`.
+        This method expects checks to already have been performed.
+        """
+        # Step 1: Keep track of the company_ids we should write on the account.
+        # We will do so only at the end, to avoid triggering the constraint that prevents duplicate codes.
+        # Writing the codes will be handled by the update of ir_property.
+        company_ids_to_write = self.company_ids
+
+        # Step 2: Check that we have write access to all the accounts and access to all the companies
+        # of these accounts.
+        self.check_access_rights('write')
+        self.check_access_rule('write')
+        if forbidden_companies := (self.sudo().company_ids - (self.env.companies & self.env.user.company_ids)):
```

Comment:
Do we need to check the current selected companies?µIsn't having access in theory enough?
Or maybe as a double check?

Reply:
It was just in order to avoid using `sudo()` to access the `code` and `company_ids` fields, which are the ones we are writing in Python. But I guess blocking the user is not great in this situation.